Adapting to HPX V1.5

hkaiser opened the pull request:
This patch adds some conditional changes that a required to make flecsi/1 compile with HPX V1.5

opensdh commented on flecsi/coloring/dcrs_utils.h:
```diff
@@ -436,7 +437,7 @@ make_dcrs_distributed(
   size_t max_global_vert_id{0};
   for(auto v : vertex_local_to_global)
-    max_global_vert_id = std::max(max_global_vert_id, v);
+    max_global_vert_id = (std::max)(max_global_vert_id, v);
```
I [hear](https://stackoverflow.com/questions/63986321/popular-header-that-misused-function-like-macro-max#comment113147406_63986321) that newer versions of Windows might not need this hack anymore.  Are we sure that we have to use it?

hkaiser replied:
The Windows headers still define min()/max() as macros. One can disable those macros by adding `-DNOMINMAX` to the compilers command line. This is certainly an option while compiling FLeCSI itself, however might interfere with dependent projects. Please advise how to proceed. I personally think that the additional parentheses are the least invasive solution, but I'm happy do adapt the code to your guidelines.

opensdh replied:
I think I agree that the parentheses, as odd as they look, are the cleanest solution for a library that shouldn't be requiring clients to globally configure their preprocessor.  I just wanted to make sure that we weren't doing so because it was _once_ necessary.

opensdh commented on flecsi/execution/test/finite_difference_dense.cc:
```diff
@@ -115,7 +115,7 @@ check_results(mesh<ro> mesh, field<ro, ro, ro> values, size_t global_target) {
 
     if(std::abs(v - t) > test_tolerance) {
-      printf("[Rank %lu] at [%lu,%lu] %.15e != %.15e\n", rank, i, j, v, t);
+      printf("[Rank %llu] at [%llu,%llu] %.15e != %.15e\n", rank, i, j, v, t);
```
If these are `size_t` the portable format code is `%zu`.

opensdh commented on flecsi/data/data_client_handle.h:
```diff
@@ -50,7 +50,7 @@ struct data_client_handle_base_u : public DATA_CLIENT_TYPE,
     UNMAPPED_PERMISSIONS,
     DATA_POLICY> & h)
-    : DATA_POLICY(h), DATA_CLIENT_TYPE(h), type_hash(h.type_hash),
+    : DATA_CLIENT_TYPE(h), DATA_POLICY(h), type_hash(h.type_hash),
```
Lots of changes like this one seem to be avoiding warnings that have nothing to do with HPX.  They do appear to be in their own commit, but it would really be nicer to have them in their own merge request (so that we can tell what's HPX-specific more easily).

hkaiser replied:
Correct, those warnings were produced by different compilers (partly MSVC, but also clang) and mostly are unrelated to the HPX backend. I will try separating those into their own merge request, however.

hkaiser replied:
See #645 for all changes unrelated to HPX.

opensdh commented on flecsi/utils/mpi_type_traits.h:
```diff
@@ -121,7 +121,7 @@ struct mpi_typetraits_u {
     else if constexpr(is_same_v<TYPE, complex<long double>>)
       return MPI_CXX_LONG_DOUBLE_COMPLEX;
-    else if constexpr(is_same_v<TYPE, byte>)
+    else if constexpr(is_same_v<TYPE, unsigned char>)
```
`unsigned char` is [already handled above](laristra/flecsi/blob/0d0f74fc2542320761b33f563193e28747031ae1/flecsi/utils/mpi_type_traits.h#L65); why would we want it to be `MPI_BYTE` instead?

hkaiser replied:
Fair point. I'll go back and check why I thought that this was needed.

hkaiser replied:
This has been changed to refer to `std::byte` instead in #645.

opensdh reviewed (commented):
I haven't gone through all(!) the changes yet, but here are some initial/procedural points.

hkaiser committed:
Changes needed to adapt for HPX V1.5

- enable all tests for HPX

hkaiser commented:
This has been completely redone (please note I force-pushed to the underlying branch). The changes unrelated to HPX have now been separated and are available as #645. This PR has now only changes that are required to have HPX V1.5 as a backend.

hkaiser commented:
Please also note that this PR depends on laristra/cinch#187

opensdh commented on flecsi/execution/hpx/execution_policy.h:
```diff
@@ -28,183 +30,250 @@
-  execute(Exec && exec, T fun, A && targs) {
-    auto user_fun = (reinterpret_cast<RETURN (*)(ARG_TUPLE)>(fun));
+  static decltype(auto) execute(Exec && exec, T function, A && targs) {
+    auto user_fun = reinterpret_cast<RETURN (*)(std::decay_t<A>)>(function);
```
Should this really be `A`, which is deduced, instead of `ARG_TUPLE`, which is presumably known from the function?

hkaiser commented on flecsi/execution/hpx/execution_policy.h:
```diff
@@ -28,183 +30,250 @@
+    //         std::move(function), task_args);
+    //     }
+    hpx_future_u<RETURN> future = executor_u<RETURN, ARG_TUPLE>::execute(
+      context_t::instance().get_default_executor(), std::move(function),
```
`function` is a scalar, `std::move` is not necessary`

hkaiser commented on flecsi/execution/hpx/execution_policy.h:
```diff
@@ -28,183 +30,250 @@
-        clog(info) << "Executing MPI task: " << KEY << std::endl;
-      }
+      return future
+        .then([&](hpx_future_u<RETURN> && future) {
```
Restrict capture to necessary values to avoid capturing things that might go out of scope.

hkaiser commented on flecsi/execution/hpx/runtime_main.cc:
```diff
@@ -39,22 +39,18 @@ main(int argc, char ** argv) {
   MPI_Init(&argc, &argv);
```
MPI needs to be initialized multi-threaded.

hkaiser committed:
Addressing review comments

hkaiser commented:
@opensdh, @ipdemes: I have addressed all review comments as discussed. Please note that I have pushed additional changes to https://github.com/laristra/cinch/pull/187.

hkaiser committed:
Making sure task arguments stay alive during task execution

# Conflicts:
#	cinch

hkaiser committed:
Adding task dependencies for HPX backend

- flyby: make FLECSI_USE_AGGCOMM=On default for HPX backend

ollielo commented on flecsi/execution/hpx/execution_policy.h:
```diff
@@ -219,9 +242,29 @@ struct FLECSI_EXPORT hpx_execution_policy_t {
-    hpx_future_u<RETURN> future = hpx::async(std::move(func));
-    future.wait();
+    // force unwrapping of returned future
+    hpx_future_u<RETURN> future = hpx::dataflow(
```
@hkaiser 

How asynchronous are hpx::dataflow and hpx::future.then? Do they just submit a "task" and its continuation and return immediately?

hkaiser replied:
> How asynchronous are hpx::dataflow and hpx::future.then? Do they just submit a "task" and its continuation and return immediately?

Yes, that's exactly what they do. They will schedule a new task once the futures they depend on have become ready.

hkaiser committed:
More work on task dependencies

- adding future to data_client_handle, integrate with task dependency logic
- flyby: optimize execution_policy::execute_task

hkaiser committed:
Work around more integral conversion warnings

hkaiser committed:
Make sure spmd_init finishes before returning

hkaiser committed:
Attempt to fix remaining test failures

andrewreisner commented on flecsi/execution/hpx/task_add_dependencies.h:
```diff
@@ -0,0 +1,202 @@
+
+    auto & h = a.handle;
+    // Skip Read Only handles
+    if constexpr((SHARED_PERMISSIONS == ro) || (GHOST_PERMISSIONS == rw) ||
```
Do the `EXCLUSIVE_PERMISSIONS` not need to be used to setup task dependencies?

hkaiser replied:
Excellent question. I essentially copied the conditions from the MPI backend, so I might have missed things. Is there an explanation of how (and which) permissions are to be interpreted to deduce dependencies? Or in general, what is encoded by the different types of permissions?

ollielo replied:
The MPI backend does not do dependency analysis. The check on permission is only for ghost copy purpose. Thus the task_[epi|pro]logue would skip handles that does not write to ghost or shared region.

For a more completed dependency analysis, the permission on Exclusive should be taken into account.

hkaiser replied:
Thanks for this explanation. I'm still unsure about the role of the various permissions, though. Would you be able to recommend some documentation or similar to read up about this?

ollielo replied:
Minor correction. `GHOST_PERMISSION == rw` is used to signify the values of ghost cells are computed locally (flecsale does that) and thus ghost exchange is not necessary. Again, I don't know what the implication to dependency analysis is.

andrewreisner replied:
https://github.com/laristra/flecsi/blob/1/flecsi/data/data.md#permissions

ollielo commented:
The probably will be better answered by Irina.

On Mon, Nov 30, 2020 at 4:24 PM Hartmut Kaiser <notifications@github.com>
wrote:

> *@hkaiser* commented on this pull request.
> ------------------------------
>
> In flecsi/execution/hpx/task_add_dependencies.h
> <https://github.com/laristra/flecsi/pull/643#discussion_r532968834>:
>
> > +                                 indices of the index partition.
> +
> +   */
> +
> +  template<typename T,
> +    size_t EXCLUSIVE_PERMISSIONS,
> +    size_t SHARED_PERMISSIONS,
> +    size_t GHOST_PERMISSIONS>
> +  void handle(dense_accessor<T,
> +    EXCLUSIVE_PERMISSIONS,
> +    SHARED_PERMISSIONS,
> +    GHOST_PERMISSIONS> & a) {
> +
> +    auto & h = a.handle;
> +    // Skip Read Only handles
> +    if constexpr((SHARED_PERMISSIONS == ro) || (GHOST_PERMISSIONS == rw) ||
>
> Thanks for this explanation. I'm still unsure about the role of the
> various permissions, though. Would you be able to recommend some
> documentation or similar to read up about this?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/laristra/flecsi/pull/643#discussion_r532968834>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFRR5JXEZI5DLICUHNNNZLSSQSTHANCNFSM4PSV3XJQ>
> .
>

hkaiser committed:
Fixing task dependencies
